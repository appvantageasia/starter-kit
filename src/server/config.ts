import chalk from 'chalk';
import { Region } from 'minio';
import { SMTPSettings } from './emails';
import { getString, getBoolean, getInteger, getNumber } from './env';

const prefix = 'APP';

const getPrefix = key => `${prefix}_${key}`;

export type RuntimeConfig = {
    version: string;

    port: number;

    publicPath: string;

    db: {
        uri: string;
        name: string;
        pool: number;
    };

    redis: {
        uri: string;
    };

    session: {
        secret: string;
        lifetime: string;
    };

    i18n: {
        locales: string[];
        defaultLocale: string;
    };

    smtp: {
        transporter: SMTPSettings;
        sender: string;
    };

    storage: {
        provider: {
            endPoint: string;
            accessKey: string;
            secretKey: string;
            useSSL?: boolean;
            port?: number;
            region: Region;
        };
        bucket: string;
    };

    sentry: {
        dsn?: string;
        release?: string;
        environment?: string;
        tracing: boolean;
        tracesSampleRate: number;
    };

    html2pdf: {
        endpoint: string;
    };
};

const getSmtpSettings = (): RuntimeConfig['smtp']['transporter'] => {
    const base = {
        host: getString(getPrefix('SMTP_HOST'), 'localhost'),
        port: getInteger(getPrefix('SMTP_PORT'), 465),
        secure: getBoolean(getPrefix('SMTP_SECURE'), false),
    };

    const user = getString(getPrefix('SMTP_USER'));

    if (!user) {
        return base;
    }

    return {
        ...base,
        auth: {
            user,
            pass: getString(getPrefix('SMTP_PASSWORD')),
        },
    };
};

const version = getString('VERSION', '0.0.0-development');

const config: RuntimeConfig = {
    version,

    port: getNumber(getPrefix('PORT'), 3000),
    publicPath: getString(getPrefix('PUBLIC_PATH'), '/public/'),

    // internationalization
    i18n: {
        locales: ['en'],
        defaultLocale: 'en',
    },

    // server runtime
    db: {
        uri: getString(getPrefix('DB_URI'), 'mongodb+srv://localhost:27017'),
        name: getString(getPrefix('DB_NAME'), 'app'),
        pool: getInteger(getPrefix('DB_POOL'), 10),
    },

    redis: {
        uri: getString(getPrefix('REDIS_URI'), 'redis://127.0.0.1:6379'),
    },

    session: {
        secret: getString(getPrefix('SESSION_SECRET')),
        lifetime: getString(getPrefix('SESSION_LIFETIME'), '1h'),
    },

    // default/system SMTP settings
    smtp: {
        transporter: getSmtpSettings(),
        sender: getString(getPrefix('SMTP_FROM'), 'noreply@appvantage.co'),
    },

    storage: {
        provider: {
            endPoint: getString(getPrefix('STORAGE_ENDPOINT')),
            accessKey: getString(getPrefix('STORAGE_ACCESS_KEY')),
            secretKey: getString(getPrefix('STORAGE_SECRET_KEY')),
            useSSL: getBoolean(getPrefix('STORAGE_SSL'), true),
            port: getInteger(getPrefix('STORAGE_PORT')),
            region: getString(getPrefix('STORAGE_REGION'), 'ap-southeast-1'),
        },
        bucket: getString(getPrefix('STORAGE_BUCKET'), 'app'),
    },

    sentry: {
        dsn: getString(getPrefix('SENTRY_DSN')),
        release: getString(getPrefix('SENTRY_RELEASE')),
        environment: getString(getPrefix('SENTRY_ENVIRONMENT')),
        tracing: getBoolean(getPrefix('SENTRY_TRACING'), true),
        tracesSampleRate: getNumber(getPrefix('SENTRY_TRACES_SAMPLE_RATE'), 1.0),
    },

    html2pdf: {
        endpoint: getString(getPrefix('HTML2PDF_ENDPOINT')),
    },
};

const exitOnError = (error: string): void => {
    console.error(error);
    process.exit(1);
};

export const runValidityChecks = () => {
    if (!config.session.secret) {
        exitOnError(chalk.red('APP_SESSION_SECRET is missing in environment variables'));
    }
};

export default config;
