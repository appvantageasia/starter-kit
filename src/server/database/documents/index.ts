export * from './User';
export * from './ExternalLink';
export * from './Setting';
export * from './Life';
