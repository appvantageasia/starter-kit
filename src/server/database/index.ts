export * from './documents';
export * from './collections';
export * from './instance';
export * from './migrate';
export * from './helpers';
