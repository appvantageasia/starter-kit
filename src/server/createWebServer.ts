// create http server
import http, { IncomingMessage, Server } from 'http';
import * as Sentry from '@sentry/node';
import { ApolloServer } from 'apollo-server-express';
import compression from 'compression';
import express, { Express } from 'express';
import { graphqlUploadExpress } from 'graphql-upload';
import config from './config';
import renderApplication from './renderApplication';
import schema from './schema';
import createContext, { Context, RootDocument } from './schema/context';
import { ApolloSentryPlugin } from './sentry';

export type WebServerCreation = {
    httpServer: Server;
    expressServer: Express;
    apolloServer: ApolloServer;
};

const createWebServer = (): WebServerCreation => {
    const apolloServer = new ApolloServer({
        schema,
        // do not use the build-in upload types
        uploads: false,

        // enable tracing for development purposes
        tracing: !!process.isDev,
        playground: !!process.isDev,

        // provide a custom context
        context: ({ req }: { req: IncomingMessage }): Promise<Context> => createContext(req),

        // provide a custom root document
        rootValue: (): RootDocument => null,

        plugins: [ApolloSentryPlugin],
    });

    const expressServer = express();
    expressServer.disable('x-powered-by');
    expressServer.use(compression());
    expressServer.use(express.json());
    expressServer.use(express.urlencoded({ extended: true }));
    expressServer.use(Sentry.Handlers.requestHandler());

    if (config.sentry.tracing) {
        expressServer.use(Sentry.Handlers.tracingHandler());
    }

    // serve static files
    expressServer.use('/public', express.static('public'));

    // serve graphql API
    expressServer.use(
        '/graphql',
        graphqlUploadExpress(),
        apolloServer.getMiddleware({ bodyParserConfig: { limit: '50mb' }, path: '/' })
    );

    // otherwise fallback on the application
    expressServer.use((req, res, next) => {
        if (req.method !== 'GET') {
            // only accept GET requests
            next();
        } else {
            renderApplication(req, res, next);
        }
    });

    // sse the sentry error handler before any other error handler
    expressServer.use(Sentry.Handlers.errorHandler());

    // then here comes our error handler
    // eslint-disable-next-line no-unused-vars
    expressServer.use((error, request, response, next) => {
        console.error(error);
        response.status('500').send('Internal error');
    });

    const httpServer = http.createServer(expressServer);

    return { httpServer, apolloServer, expressServer };
};

export default createWebServer;
