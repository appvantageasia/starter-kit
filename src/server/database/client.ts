import { MongoClient, MongoClientOptions } from 'mongodb';
import config from '../core/config';
import { getEncryptionKeyId } from './helpers';
import getKMS from './kms';
import buildSchemaMap from './schemaMap';

export const getRegularOptions = (): MongoClientOptions => ({
    maxPoolSize: config.db.pool,
});

export const getRegularClient = () => MongoClient.connect(config.db.uri, getRegularOptions());

export const getEncryptedClient = async (kms: Exclude<ReturnType<typeof getKMS>, null>) => {
    const keyId = await getEncryptionKeyId();

    if (!keyId) {
        throw new Error('No encryption key ID have been provided for the schema');
    }

    return MongoClient.connect(config.db.uri, {
        ...getRegularOptions(),

        // enable command monitoring for the client
        monitorCommands: true,

        autoEncryption: {
            keyVaultNamespace: kms.keyVaultNamespace,
            kmsProviders: kms.kmsProviders,
            schemaMap: buildSchemaMap(keyId),

            // configure cryptd client
            extraOptions: {
                mongocryptdURI: config.db.cryptd.uri,
                mongocryptdBypassSpawn: config.db.cryptd.mongocryptdBypassSpawn,
                mongocryptdSpawnArgs: config.db.cryptd.mongocryptdSpawnArgs,
            },
        },
    });
};
