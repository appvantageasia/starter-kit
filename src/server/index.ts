import { program } from 'commander';
import {
    composeCommand,
    executeDataMigration,
    startServerCommand,
    startWorkerCommand,
    startBullBoardCommand,
    executeSetupMasterKey,
} from './core/commands';
import config, { runValidityChecks } from './core/config';
import createWebServer from './core/createWebServer';
import { HealthStatusManager, createHealthServer } from './core/health';
import { startPrometheusServer } from './core/prometheus';
import { listPendingMigrations } from './database';
import { setup as startWorker } from './queues';

runValidityChecks();

program.version(config.version);

program
    .command('migrate')
    .description('Execute data migration')
    .action(() => executeDataMigration(true));

program
    .command('worker')
    .description('Start worker')
    .action(() => {
        const manager = new HealthStatusManager();
        createHealthServer(manager);
        composeCommand(startWorkerCommand(manager));
    });

program
    .command('serve')
    .description('Start web server')
    .action(async () => {
        const manager = new HealthStatusManager();
        createHealthServer(manager);
        composeCommand(await startServerCommand(manager), await startPrometheusServer());
    });

program
    .command('bullBoard')
    .description('Start bull board')
    .action(async () => {
        const manager = new HealthStatusManager();
        createHealthServer(manager);
        composeCommand(startBullBoardCommand(manager));
    });

program.command('setupMasterKey').description('Setup master key for MongoDB CSFLE').action(executeSetupMasterKey);

if (process.isCLI) {
    program.parse();
}

const bundle = { createWebServer, startWorker, executeDataMigration, listPendingMigrations };

export default bundle;

export type Bundle = typeof bundle;
