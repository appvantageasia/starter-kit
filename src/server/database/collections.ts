import { Collection } from 'mongodb';
import * as documents from './documents';
import { DatabaseContext } from './instance';

export type Collections = {
    users: Collection<documents.User>;
    userSessions: Collection<documents.UserSession>;
    userWebCredentials: Collection<documents.UserWebCredential>;
    externalLinks: Collection<documents.ExternalLink>;
    settings: Collection<documents.Setting>;
    lives: Collection<documents.Life>;
};

export const getCollections = ({
    regular,
    encrypted,
}: Pick<DatabaseContext, 'regular' | 'encrypted'>): Collections => ({
    users: regular.db.collection<documents.User>('users'),
    userSessions: regular.db.collection<documents.UserSession>('userSessions'),
    userWebCredentials: regular.db.collection<documents.UserWebCredential>('userWebCredentials'),
    settings: encrypted.db.collection<documents.Setting>('settings'),
    externalLinks: regular.db.collection<documents.ExternalLink>('externalLinks'),
    lives: regular.db.collection<documents.Life>('lives'),
});
