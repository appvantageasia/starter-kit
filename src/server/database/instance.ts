import { MongoClient, Db } from 'mongodb';
import config from '../core/config';
import { getRegularClient, getEncryptedClient } from './client';
import { Collections, getCollections } from './collections';
import getKMS from './kms';

export type DatabaseContext = {
    regular: { client: MongoClient; db: Db };
    encrypted: { client: MongoClient; db: Db };
    collections: Collections;
};

/**
 * Global is used here to maintain a cached connection across hot reloads
 * in development. This prevents connections growing exponentially
 * during API Route usage.
 */

if (!global.mongo) {
    // update the global store
    global.mongo = { context: null, promise: null };
}

let cached: DatabaseContext = null;

export const getDatabaseContext = async (): Promise<DatabaseContext> => {
    if (cached && global.mongo.context) {
        return cached;
    }

    if (!global.mongo.promise) {
        const init = async (): Promise<Pick<DatabaseContext, 'regular' | 'encrypted'>> => {
            // get regular client first
            const regularClient = await getRegularClient();
            const regularDb = regularClient.db(config.db.name);

            // get encrypted client
            const kms = getKMS();
            const encryptedClient = kms ? await getEncryptedClient(kms) : regularClient;
            const encryptedDb = encryptedClient.db(config.db.name);

            return {
                regular: { client: regularClient, db: regularDb },
                encrypted: { client: encryptedClient, db: encryptedDb },
            };
        };

        // get the promise
        global.mongo.promise = init();
    }

    // wait for it
    // and assigned it globally
    global.mongo.context = await global.mongo.promise;

    // update the cache
    cached = {
        ...global.mongo.context,
        collections: getCollections(global.mongo.context),
    };

    // finally return the cache
    return cached;
};
