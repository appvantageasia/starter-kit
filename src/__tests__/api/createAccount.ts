import { gql, ApolloError } from '@apollo/client';
import { compare } from 'bcryptjs';
import { ObjectId } from 'mongodb';
import { getDatabaseContext } from '../../server/database';
import {
    composeHandlers,
    loadFixtures,
    setupDatabase,
    cleanDatabase,
    setupWebService,
    getApolloClient,
} from '../helpers';
import fixtures from './authenticate.fixture.json';

const mutation = gql`
    mutation test($username: String!, $password: String!, $email: String!) {
        createAccount(username: $username, password: $password, email: $email) {
            id
        }
    }
`;

const webService = setupWebService();

beforeEach(composeHandlers(setupDatabase, loadFixtures(fixtures), webService.initialize));

afterEach(composeHandlers(cleanDatabase, webService.cleanUp));

test('Create account rejects weak passwords', async () => {
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'y', email: 'test@test.fr' };
    const promise = client.mutate({ mutation, variables });
    await expect(promise).rejects.toBeInstanceOf(ApolloError);
    const error: ApolloError = await promise.catch(error => error);
    expect(error.graphQLErrors).toMatchSnapshot();
});

test('Create account rejects on duplicate username', async () => {
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'super0923582357word', email: 'test@test.fr' };
    const promise = client.mutate({ mutation, variables });
    await expect(promise).rejects.toBeInstanceOf(ApolloError);
    const error: ApolloError = await promise.catch(error => error);
    expect(error.graphQLErrors).toMatchSnapshot();
});

test('Create account successfully create a new user on valid inputs', async () => {
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'newUser', password: 'super0923582357word', email: 'test@test.fr' };
    const { data } = await client.mutate({ mutation, variables });
    const userId = new ObjectId(data.createAccount.id);
    const { collections } = await getDatabaseContext();
    const user = await collections.users.findOne({ _id: userId });
    expect(user).not.toBeNull();
    expect(user.username).toBe(variables.username);
    expect(user.displayName).toBe(variables.username);
    await expect(compare(variables.password, user.password)).resolves.toBe(true);
});
