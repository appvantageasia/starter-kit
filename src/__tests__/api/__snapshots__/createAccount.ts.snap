// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create account rejects on duplicate username 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "password": "Username already taken",
    },
    "locations": [
      {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "bad request",
    "path": [
      "createAccount",
    ],
  },
]
`;

exports[`Create account rejects weak passwords 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "password": "Password too weak",
    },
    "locations": [
      {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "bad request",
    "path": [
      "createAccount",
    ],
  },
]
`;
